subserver_permissions.go

```diff
@@ -41,53 +41,71 @@ var (
 	}
 )
 
-// getSubserverPermissions returns a merged map of all subserver macaroon
-// permissions.
-func getSubserverPermissions() map[string][]bakery.Op {
-	mapSize := len(faraday.RequiredPermissions) +
-		len(loop.RequiredPermissions) + len(pool.RequiredPermissions)
-	result := make(map[string][]bakery.Op, mapSize)
-	for key, value := range faraday.RequiredPermissions {
-		result[key] = value
-	}
-	for key, value := range loop.RequiredPermissions {
-		result[key] = value
-	}
-	for key, value := range pool.RequiredPermissions {
-		result[key] = value
-	}
-	for key, value := range litPermissions {
-		result[key] = value
-	}
-	return result
+// subServerName is a name used to identify a particular Lit sub-server.
+type subServerName string
+
+const (
+	poolPerms    subServerName = "pool"
+	loopPerms    subServerName = "loop"
+	faradayPerms subServerName = "faraday"
+	litPerms     subServerName = "lit"
+	lndPerms     subServerName = "lnd"
+)
+
+// PermissionsManager manages the permission lists that Lit requires.
+type PermissionsManager struct {
+	// fixedPerms is constructed once on creation of the PermissionsManager.
+	// It contains all the permissions that will not change throughout the
+	// lifetime of the manager. It maps sub-server name to uri to permission
+	// operations.
+	fixedPerms map[subServerName]map[string][]bakery.Op
+
+	// perms is a map containing all permissions that the manager knows
+	// are available for use.
+	perms map[string][]bakery.Op
 }
 
-// getAllMethodPermissions returns a merged map of lnd's and all subservers'
-// method macaroon permissions.
-func getAllMethodPermissions() map[string][]bakery.Op {
-	subserverPermissions := getSubserverPermissions()
-	lndPermissions := lnd.MainRPCServerPermissions()
-	mapSize := len(subserverPermissions) + len(lndPermissions) +
-		len(whiteListedMethods)
-	result := make(map[string][]bakery.Op, mapSize)
-	for key, value := range lndPermissions {
-		result[key] = value
+// NewPermissionsManager constructs a new PermissionsManager instance and
+// collects any of the fixed permissions.
+func NewPermissionsManager() (*PermissionsManager, error) {
+	permissions := make(map[subServerName]map[string][]bakery.Op)
+	permissions[faradayPerms] = faraday.RequiredPermissions
+	permissions[loopPerms] = loop.RequiredPermissions
+	permissions[poolPerms] = pool.RequiredPermissions
+	permissions[litPerms] = litPermissions
+	permissions[lndPerms] = lnd.MainRPCServerPermissions()
+	for k, v := range whiteListedLNDMethods {
+		permissions[lndPerms][k] = v
 	}
-	for key, value := range subserverPermissions {
-		result[key] = value
-	}
-	for key, value := range whiteListedMethods {
-		result[key] = value
+
+	allPerms := make(map[string][]bakery.Op)
+	for _, perms := range permissions {
+		for k, v := range perms {
+			allPerms[k] = v
+		}
 	}
-	return result
+
+	return &PermissionsManager{
+		fixedPerms: permissions,
+		perms:      allPerms,
```

Comment:
Isn't setting this on start a contradiction to the godoc? Meaning that if we call `GetPermissionsList()` directly after creating the manager, we'd always get all permissions?
Changing this probably means needing to call `OnLNDBuildTags()` in the itest manually (for baking the super macaroon).

Reply:
well, we never _remove_ any perms from this map. We only ever add more perms once we get the build tags from lnd.

Reply:
Meaning that `GetPermissionsList` doesnt ever provide you with false positives.

Reply:
so `perms` represents all permissions that we know we definitely  have

Reply:
the godoc for the `perms` member is also updated in the next commit